cs-mail/src/main/java/io/cloudslang/content/mail/services/GetMailAttachmentService.java

```diff
@@ -0,0 +1,305 @@
+
+
+    private String downloadAttachment(Message message, String attachment, String characterSet, String path, boolean overwrite) throws Exception {
+        if (!message.isMimeType(MimeTypes.TEXT_PLAIN) && !message.isMimeType(MimeTypes.TEXT_HTML)) {
+
+            Multipart mpart = (Multipart) message.getContent();
+            File f = new File(path);
+            if (f.isDirectory()) {
+                if (mpart != null) {
+                    for (int i = 0, n = mpart.getCount(); i < n; i++) {
+                        Part part = mpart.getBodyPart(i);
+
+                        if (input.isEncryptedMessage() && part.getContentType() != null &&
+                                part.getContentType().equals(SecurityConstants.ENCRYPTED_CONTENT_TYPE)) {
+                            part = decryptPart((MimeBodyPart) part);
+                        }
+
+                        if (part != null && part.getFileName() != null &&
+                                (MimeUtility.decodeText(part.getFileName()).equalsIgnoreCase(attachment) ||
+                                        part.getFileName().equalsIgnoreCase(attachment))
+                        ) {
+                            String disposition = part.getDisposition();
+                            if (disposition != null && disposition.equalsIgnoreCase(Part.ATTACHMENT)) {
+                                String partPrefix = part.getContentType().toLowerCase();
+                                if (partPrefix.startsWith(io.cloudslang.content.mail.constants.MimeTypes.TEXT_PLAIN)) {
+                                    String attachmentContent = attachmentToString(part, characterSet);
+                                    writeTextToNewFile(path + "/" + attachment, attachmentContent, overwrite);
+                                    results.put(io.cloudslang.content.constants.OutputNames.RETURN_RESULT, MimeUtility.decodeText(attachmentContent));
+                                } else {
+                                    writeNewFile(path + "/" + attachment, part, overwrite);
+                                }
+                            }
+                            return MimeUtility.decodeText(part.getFileName());
+                        }
+                    }
+                }
+            } else throw new Exception("The Folder " + path + " does not exist.");
+        }
+        throw new Exception("The email does not have an attached file by the name " + attachment + ".");
+    }
+
+
+    private static String writeNewFile(String path, Part part, boolean overwrite) throws Exception {
+
+        InputStream is = null;
+        OutputStream os = null;
+
+        try {
+            is = part.getInputStream();
+
+            File f = new File(path);
+            if (!f.exists() || overwrite) {
+                if (!f.createNewFile()) {
+                    throw new IOException("Could not create file at path: " + path);
+                }
+
+                os = new FileOutputStream(f);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *os*. The resource is referenced in statements at the following lines: 255, 270, 274, 280. The resource closure statement is at line: 281. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by InputStream.close. Either a) close *os* in a try-finally block or b) close the resource by declaring *os* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).

Reply:
false positive